k2/csrc/context.cu

```diff
@@ -101,7 +102,7 @@ class CudaContext : public Context {
 ContextPtr GetCpuContext() { return std::make_shared<CpuContext>(); }
 
 ContextPtr GetCudaContext(int32_t gpu_id /*= -1*/) {
```

Comment:
I would like to leave `GetCudaContext` as before, i.e. returns a `raw` GPU context (and `GetCpuContext` returns a `raw` CPU Context). While returning `PyTorchContext` from a function like `GetPyTorchContext`, and as you said, the `PyTorchContext` could be on either GPU device or CPU device. We definitely will create other Context, e.g. `TensorFlowContext`.

Reply:
My original plan was like the following:

```cpp
// context.cu
Leave as it is
```

```cpp
// PytorchContext.h
// like the current one

// PytorchContext.cu
ContextPtr GetCudaContext(int32_t gpu_id /*= -1*/) {
  return std::make_shared<PytorchContext>(gpu_id);
}
```

There would be an option in `CMakeLists.txt` to configure whether `PytorchContext.cu` **OR** `context.cu` should
be compiled.

To support other frameworks, e.g., tensorflow, we can create `TensorflowContext.h` and `TensorflowContext.cu`.

One of the advantages of this approach is that we can use `GetCudaContext`  in the code without worrying about
where it comes from. The drawback is that it supports only one framework.